### PR TITLE
Issue 877 - core-metadata DeviceService: Eliminate Mongo Types from Models (Refactored)

### PR DESCRIPTION
**Pull request by michaelestrin**
Issue #877 - core-metadata DeviceService: Eliminate Mongo Types from Models (Refactored)

* Eliminated unnecessary mongoDeviceService struct and moved SetBSON/GetBSON into DeviceService model.

* Separated responsibilities -- Mongo model updated to represent data as it exists in Mongo (including DBRefs) and conversion (ToContract/FromContract) now responsible for transform to/from DBRefs.

* Added addressableTransform contract to facilitate conversion of addressable to/from DBRef within individual model's ToContract/FromContract.

Signed-off-by: Michael Estrin <m.estrin@dell.com>

### CODE REVIEW

File: internal/pkg/db/mongo/models/addressable.go [hunk 1]
```diff
@@ -37,7 +43,7 @@ type Addressable struct {
 	Origin     int64         `bson:"origin"`
 }
 
-func (a Addressable) ToContract() contract.Addressable {
+func (a *Addressable) ToContract() contract.Addressable {
```
Comment (tsconn23) [inline internal/pkg/db/mongo/models/addressable.go]: I don't think the receiver needs to be a pointer here. No internal state is mutated. Small potatoes.

File: internal/pkg/db/mongo/models/event.go [hunk 2]
```diff
@@ -34,7 +34,7 @@ type Event struct {
 	dbRefs   []mgo.DBRef
 }
 
-func (e Event) ToContract() contract.Event {
+func (e *Event) ToContract() contract.Event {
```
Comment (tsconn23) [inline internal/pkg/db/mongo/models/event.go]: Again, I don't understand why some of these receivers were made into pointers.

**Comment by tsconn23** [comment]
Overall LGTM. Will leave as is for others who may be interested to review. Would like to know about the rationale for the new pointer receivers. It's a small point, but would like to understand.

**Comment by jdharms** [comment]
> Overall LGTM. Will leave as is for others who may be interested to review. Would like to know about the rationale for the new pointer receivers. It's a small point, but would like to understand.

From The Go Programming Language (book):

> In a realistic program, convention dictates that if any method of Point has a pointer receiver, then all methods of Point should have a pointer receiver, even ones that don’t strictly need it.

Where "Point" is a type they're writing for illustrative purposes in the book.

I think this is @michaelestrin 's reasoning here as well.

**Comment by michaelestrin** [comment]
@jdharms beat me to it.  My response was in flight when his showed up.  That's exactly it.

**Comment by tsconn23** [comment]
@jdharms @michaelestrin You missed a couple spots then.

```
func (e Event) GetBSON() (interface{}, error) 
func (e Event) GetDBRefs() []mgo.DBRef 
```

**Comment by michaelestrin** [comment]
> @jdharms @michaelestrin You missed a couple spots then.
> 
> ```
> func (e Event) GetBSON() (interface{}, error) 
> func (e Event) GetDBRefs() []mgo.DBRef 
> ```

Fixed.

**Pull request merged**